src/de.hpi.swa.graal.squeak/src/de/hpi/swa/graal/squeak/nodes/bytecodes/ReturnBytecodes.java

```diff
@@ -67,9 +75,33 @@ protected final Object doNonLocalReturn(final VirtualFrame frame) {
         }
 
         @Specialization(guards = {"isCompiledBlockObject(code)"})
-        protected final Object doClosureReturn(final VirtualFrame frame) {
+        protected final Object doClosureReturnFromMaterialized(final VirtualFrame frame) {
             // Target is sender of closure's home context.
-            throw new NonLocalReturn(getReturnValue(frame), FrameAccess.getClosure(frame).getHomeContext().getFrameSender());
+            final ContextObject homeContext = FrameAccess.getClosure(frame).getHomeContext();
+            assert homeContext.getProcess() != null;
+            final boolean homeContextNotOnTheStack = homeContext.getProcess() != code.image.getActiveProcess(readNode);
+            final Object caller = homeContext.getFrameSender();
+            if (caller == NilObject.SINGLETON || homeContextNotOnTheStack) {
+                getCannotReturnNode().executeSend(frame, getGetOrCreateContextNode().executeGet(frame), getReturnValue(frame));
```

Comment:
First, I  thought this needs a `BranchProfile`, but `getCannotReturnNode()` basically acts like a `BranchProfile` here.